Send data for browser.storage to the UI process

kiaraarose opened the pull request:
#### 0e52b5dc8c72b2d398724308508dc149ac4b6c39
<pre>
Send data for browser.storage to the UI process
<a href="https://bugs.webkit.org/show_bug.cgi?id=267782">https://bugs.webkit.org/show_bug.cgi?id=267782</a>
<a href="https://rdar.apple.com/121279041">rdar://121279041</a>

Reviewed by Timothy Hatcher.

This patch sends the data to be added or removed from storage over to the UI process. Next and
final steps for the Storage API is to adopt the _WKWebExtensionSQLiteStore to manage the data added
with this API. This is being tracked in <a href="https://webkit.org/b/267649.">https://webkit.org/b/267649.</a>

In addition, this patch adds support for browser.session.setAccessLevel. To do this, add a boolean
representing the access level to WebExtensionContextParameters so that the value can be read when
determining to expose the &apos;session&apos; API.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_cannot_be_forward_declared):
* Source/WebKit/Shared/Extensions/WebExtensionConstants.h:
* Source/WebKit/Shared/Extensions/WebExtensionContext.serialization.in:
* Source/WebKit/Shared/Extensions/WebExtensionContextParameters.h:
* Source/WebKit/Shared/Extensions/WebExtensionContextParameters.serialization.in:
* Source/WebKit/Shared/Extensions/WebExtensionStorageType.h:
Copied from Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIStorage.h.
(WebKit::toAPIPrefixString):
* Source/WebKit/Shared/Extensions/WebExtensionStorageType.serialization.in:
Copied from Source/WebKit/Shared/Extensions/WebExtensionContext.serialization.in.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIStorageCocoa.mm: Added.
(WebKit::WebExtensionContext::storageGet):
(WebKit::WebExtensionContext::storageGetBytesInUse):
(WebKit::WebExtensionContext::storageSet):
(WebKit::WebExtensionContext::storageRemove):
(WebKit::WebExtensionContext::storageClear):
(WebKit::WebExtensionContext::storageSetAccessLevel):
(WebKit::WebExtensionContext::isValidStorageRequest):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::load):
(WebKit::WebExtensionContext::getTab):
(WebKit::WebExtensionContext::isBackgroundWebView):
(WebKit::WebExtensionContext::setSessionStorageAllowedInContentScripts):
Set the access level and send a message to update the value on the WebExtensionContextParameters.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.cpp:
(WebKit::WebExtensionContext::parameters const):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
(WebKit::WebExtensionContext::isSessionStorageAllowedInContentScripts const):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIStorageAreaCocoa.mm:
(WebKit::WebExtensionAPIStorageArea::isPropertyAllowed):
(WebKit::WebExtensionAPIStorageArea::get):
(WebKit::WebExtensionAPIStorageArea::getBytesInUse):
(WebKit::WebExtensionAPIStorageArea::set):
(WebKit::WebExtensionAPIStorageArea::remove):
(WebKit::WebExtensionAPIStorageArea::clear):
(WebKit::WebExtensionAPIStorageArea::setAccessLevel):
(WebKit::WebExtensionAPIStorageArea::onChanged):
(WebKit::WebExtensionAPIStorageArea::quotaBytes):
(WebKit::WebExtensionAPIStorageArea::quotaBytesPerItem):
(WebKit::WebExtensionAPIStorageArea::maxItems):
(WebKit::WebExtensionAPIStorageArea::maxWriteOperationsPerHour):
(WebKit::WebExtensionAPIStorageArea::maxWriteOperationsPerMinute):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIStorageCocoa.mm:
(WebKit::WebExtensionAPIStorage::isPropertyAllowed):
(WebKit::WebExtensionAPIStorage::local):
(WebKit::WebExtensionAPIStorage::session):
(WebKit::WebExtensionAPIStorage::sync):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIStorage.h:
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIStorageArea.h:
(WebKit::WebExtensionAPIStorageArea::WebExtensionAPIStorageArea):
* Source/WebKit/WebProcess/Extensions/Cocoa/WebExtensionContextProxyCocoa.mm:
(WebKit::WebExtensionContextProxy::getOrCreate):
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIStorage.idl:
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIStorageArea.idl:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.cpp:
(WebKit::WebExtensionContextProxy::setStorageAccessLevel):
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.h:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIStorage.mm:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/273286@main">https://commits.webkit.org/273286@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2783b5ebc1598a49d7e8a5d425c16e3037fe42d9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [  ~~🧪 style~~](https://ews-build.webkit.org/#/builders/38/builds/34994 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/13880 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/37067 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/37739 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/31605 "Built successfully") 
| [  ~~🧪 bindings~~](https://ews-build.webkit.org/#/builders/9/builds/36152 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/16268 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/10958 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/30543 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/11/builds/35536 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/11761 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/31205 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/10307 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🧪 webkitpy~~](https://ews-build.webkit.org/#/builders/6/builds/35262 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/10367 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/31294 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/38991 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/31825 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/31623 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/36392 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/10472 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/8392 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/34378 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/12273 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/11004 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4497 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/11336 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionStorageType.h:
```diff
@@ -0,0 +1,70 @@
+    WebKit::WebExtensionStorageType,
+    WebKit::WebExtensionStorageType::Local,
+    WebKit::WebExtensionStorageType::Session,
+    WebKit::WebExtensionStorageType::Sync
```
Indent one more level.

kiaraarose commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIStorage.mm:
```diff
@@ -46,13 +47,18 @@
-        @"browser.test.assertThrows(() => browser.storage.local.get(Date.now()), /an invalid parameter was passed/i)",
+        @"browser.test.assertThrows(() => browser.storage.local.get(Date.now()), /'items' value is invalid, because an object or a string or an array of strings or null is expected, but a number was provided/i)",
 
 
```
remove extra newline

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIStorageCocoa.mm:
```diff
@@ -0,0 +1,143 @@
+        return false;
+    }
+
+    return true;
```
This will fail for popup pages. I think the check can be simplified to focus on tab pages. If it isn't a tab page, just allow it since it is likely the background or a popup.

Also having the error be "tab" related is not correct here, since none of the storage APIs take tab ids. The check is about if they have access or not. The extensions can't get here if it does not have access to the tab, otherwise they couldn't have injected a script using `storage`, so I'm not sure we need a check here at all.

If we do need a check, it would be more like this:

```
    RefPtr tab = getTab(webPageProxyIdentifier, std::nullopt, WebExtensionContext::IgnoreExtensionAccess::No);
    if (!tab)
        return true;

    if (!tab->extensionHasAccess() || !tab->extensionHasPermission()) {
        *errorMessage = @"storage access not allowed";
        return false;
    }

    return true;
```

xeenon replied:
Revised after talking about this:

```
RefPtr page = WebProcessProxy::webPage(webPageProxyIdentifier);
if (page && isURLForThisExtension(page->pageLoadState().activeURL()))
    return true;

RefPtr tab = getTab(webPageProxyIdentifier);
if (!tab || !tab->extensionHasPermission()) {
    *errorMessage = @"storage access not allowed";
    return false;
}

return true;
```

pxlcoder commented on Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIStorageArea.h:
```diff
@@ -34,43 +34,40 @@
-    void setAccessLevel(NSDictionary *, Ref<WebExtensionCallbackHandler>&&, NSString **outExceptionString);
-
-    WebExtensionAPIEvent& onChanged();
+    void get(WebPage *, id items, Ref<WebExtensionCallbackHandler>&&, NSString **outExceptionString);
```
Since `WebPage` is a C++ type, the `*` should be next to the type. There are also some instances in the implementation files in this patch that need to be corrected.

```suggestion
    void get(WebPage*, id items, Ref<WebExtensionCallbackHandler>&&, NSString **outExceptionString);
```

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionConstants.h:
```diff
@@ -35,6 +35,9 @@ static constexpr size_t webExtensionDeclarativeNetRequestMaximumNumberOfEnabledR
 
 // MARK: Storage constants.
+static constexpr NSString *accessLevelTrustedContexts = @"TRUSTED_CONTEXTS";
+static constexpr NSString *accessLevelTrustedAndUntrustedContexts = @"TRUSTED_AND_UNTRUSTED_CONTEXTS";
```
`constexpr` does not do anything for ObjC types.

Keeping these in `WebExtensionAPIStorageAreaCocoa.mm` still makes the most sense, as they are only used in that file.

kiaraarose replied:
I use `accessLevelTrustedAndUntrustedContexts` in WebExtensionContextAPICocoa which is why I moved them over to the constants file. Will change to `static NSString * const` though.

xeenon replied:
Oh I see now. We should make a new enum for that. We don't want the JS API values to leak into the UI process implementation. Sending an enum is easier on the IPC too (a small number, vs a string that needs encoded).

kiaraarose replied:
Good point! Will do.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -3192,6 +3219,17 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+
+    writeStateToStorage();
+
+    extensionController()->sendToAllProcesses(Messages::WebExtensionContextProxy::SetStorageAccessLevel(allowed), identifier());
```
This should check `isLoaded()` and return early if false. Currently this is only called when loaded, but it will guard the case that someone might call it when not loaded, when `extensionController` is null.

```
if (!isLoaded())
    return false;
        
extensionController()->sendToAllProcesses(Messages::WebExtensionContextProxy::SetStorageAccessLevel(allowed), identifier());
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIStorageCocoa.mm:
```diff
@@ -36,12 +36,19 @@
+{
+    ASSERT(propertyName == "session"_s);
+
+    return extensionContext().isSessionStorageAllowedInContentScripts() || isForMainWorld();
```
This should likely be:

```
if (propertyName == "session"_s)
    return extensionContext().isSessionStorageAllowedInContentScripts() || isForMainWorld();

ASSERT_NOT_REACHED();
return false;
```

To be future proof.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIStorageArea.h:
```diff
@@ -34,43 +34,40 @@
     }
 
-    StorageType m_type;
+    WebExtensionStorageType m_type;
```
Add default initializer.

```
WebExtensionStorageType m_type { WebExtensionStorageType::Local };
```

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIStorage.mm:
```diff
@@ -66,12 +71,123 @@
+        @"browser.test.assertEq(typeof browser.storage.session.MAX_WRITE_OPERATIONS_PER_HOUR, 'undefined')",
+
+        @"browser.test.assertEq(typeof browser.storage.local.MAX_WRITE_OPERATIONS_PER_MINUTE, 'undefined')",
+        @"browser.test.assertEq(typeof browser.storage.session.MAX_WRITE_OPERATIONS_PER_MINUTE, 'undefined')",
```
This is fine as-is. But could be:

```
browser.test.assertEq(browser.storage.local.setAccessLevel, undefined)
```

Also we should start using [optional chaining](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining) for property access. That will prevent the test from throwing an exception if any of the properties in the "chain" are null / undefined. Especially since an exception causes our tests to timeout currently.

```
browser.test.assertEq(browser?.storage?.local?.setAccessLevel, undefined)
```

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIStorage.mm:
```diff
@@ -66,12 +71,123 @@
         @"browser.test.assertThrows(() => browser.storage.local.remove([1]), /'keys' value is invalid, because a string or an array of strings is expected, but an array of other values was provided/i)",
 
+        @"browser.test.assertThrows(() => browser.storage.session.setAccessLevel('INVALID_ACCESS_LEVEL'), /'accessOptions' value is invalid, because an object is expected/i)",
+        @"browser.test.assertThrows(() => browser.storage.session.setAccessLevel({ 'accessLevel': 'INVALID_ACCESS_LEVEL' }), /'accessLevel' value is invalid, because it must specify either 'TRUSTED_CONTEXTS' or 'TRUSTED_AND_UNTRUSTED_CONTEXTS'/i)",
```
`'accessLevel'` does not need quotes.

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIStorage.mm:
```diff
@@ -66,12 +71,123 @@
+
+        @"  await browser.storage.session.setAccessLevel({ accessLevel: 'TRUSTED_CONTEXTS' })",
+
+        @"  const response = await browser.test.assertSafeResolve(() => browser.tabs.sendMessage(tabId, { content: 'Access level set' }))",
```
Can just be `'Access level set'` instead of `{ content: 'Access level set' }`

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIStorage.mm:
```diff
@@ -66,12 +71,123 @@
+
+    auto *contentScript = Util::constructScript(@[
+        @"browser.runtime.onMessage.addListener((message, sender, sendResponse) => {",
+        @"  browser.test.assertEq(message.content, 'Access level set', 'Should receive the correct message content')",
```
Just `message` if you make the other change in the `sendMessage` part. Otherwise `message?.content` to avoid breaking if the message is a different object.

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIStorage.mm:
```diff
@@ -66,12 +71,123 @@
+
+        @"  await browser.storage.session.setAccessLevel({ accessLevel: 'TRUSTED_AND_UNTRUSTED_CONTEXTS' })",
+
+        @"  const response = await browser.test.assertSafeResolve(() => browser.tabs.sendMessage(tabId, { content: 'Access level set' }))",
```
Ditto.

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIStorage.mm:
```diff
@@ -66,12 +71,123 @@
+
+    auto *contentScript = Util::constructScript(@[
+        @"browser.runtime.onMessage.addListener((message, sender, sendResponse) => {",
+        @"  browser.test.assertEq(message.content, 'Access level set', 'Should receive the correct message content')",
```
Ditto.

xeenon commented on Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIStorageArea.idl:
```diff
@@ -28,17 +28,16 @@
+    [RaisesException, NeedsPage] void getBytesInUse([Optional, NSObject, DOMString] any keys, [Optional, CallbackHandler] function callback);
+    [RaisesException, NeedsPage] void set([NSDictionary] any items, [Optional, CallbackHandler] function callback);
+    [RaisesException, NeedsPage] void remove([NSObject] any keys, [Optional, CallbackHandler] function callback);
+    [NeedsPage] void clear([Optional, CallbackHandler] function callback);
```
Can use `NeedsPageWithCallbackHandler` on the `interface` above.

xeenon commented on Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIStorageArea.idl:
```diff
@@ -28,17 +28,16 @@
     ReturnsPromiseWhenCallbackIsOmitted,
```
Since all the calls need the page now, you can add `NeedsPageWithCallbackHandler` here.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIStorageAreaCocoa.mm:
```diff
@@ -211,17 +219,31 @@
+
+    if (!m_onChanged)
+        m_onChanged = WebExtensionAPIEvent::create(forMainWorld(), runtime(), extensionContext(), WebExtensionEventListenerType::StorageOnChanged);
+    return *m_onChanged;
```
Add a newline above.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIStorageAreaCocoa.mm:
```diff
@@ -83,8 +85,12 @@
+                callback->reportError(error.value());
+            else
+                callback->call(parseJSON(API::Data::create(dataReference.value())));
+        }, extensionContext().identifier().toUInt64());
```
`.toUInt64()` is not needed here. I mistakenly had it on some others.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIStorageAreaCocoa.mm:
```diff
@@ -136,11 +137,15 @@
+            callback->reportError(error.value());
+        else
+            callback->call(@(size.value()));
+    }, extensionContext().identifier().toUInt64());
```
Remove `.toUInt64()`.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIStorageAreaCocoa.mm:
```diff
@@ -160,11 +165,15 @@
+            callback->reportError(error.value());
+        else
+            callback->call();
+    }, extensionContext().identifier().toUInt64());
```
Remove `.toUInt64()`.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIStorageAreaCocoa.mm:
```diff
@@ -173,26 +182,25 @@
+            callback->reportError(error.value());
+        else
+            callback->call();
+    }, extensionContext().identifier().toUInt64());
```
Remove `.toUInt64()`.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIStorageAreaCocoa.mm:
```diff
@@ -173,26 +182,25 @@
+            callback->reportError(error.value());
+        else
+            callback->call();
+    }, extensionContext().identifier().toUInt64());
```
Remove `.toUInt64()`.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIStorageAreaCocoa.mm:
```diff
@@ -211,17 +219,31 @@
+            callback->reportError(error.value());
+        else
+            callback->call();
+    }, extensionContext().identifier().toUInt64());
```
Remove `.toUInt64()`.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIStorageAreaCocoa.mm:
```diff
@@ -44,29 +46,29 @@
 static NSString * const accessLevelKey = @"accessLevel";
 
-static NSString * const accessLevelTrustedContexts = @"TRUSTED_CONTEXTS";
-static NSString * const accessLevelTrustedAndUntrustedContexts = @"TRUSTED_AND_UNTRUSTED_CONTEXTS";
```
Keep here.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -85,6 +87,8 @@
 static NSString * const lastSeenVersionStateKey = @"LastSeenVersion";
 static NSString * const lastLoadedDeclarativeNetRequestHashStateKey = @"LastLoadedDeclarativeNetRequestHash";
 
+static NSString * const sessionStorageAllowedInContentScriptsKey = @"sessionStorageAllowedInContentScripts";
```
`@"SessionStorageAllowedInContentScripts"`

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionContextParameters.h:
```diff
@@ -47,6 +47,7 @@ struct WebExtensionContextParameters {
 
     double manifestVersion;
     bool testingMode;
+    bool isSessionStorageAllowedInContentScripts;
```
These all should have default initializers, so they don't get random memory if they are not explicitly set.

```
double manifestVersion { 0 };
bool testingMode { false };
bool isSessionStorageAllowedInContentScripts { false };
```

kiaraarose committed:
Send data for browser.storage to the UI process
https://bugs.webkit.org/show_bug.cgi?id=267782
rdar://121279041

Reviewed by Timothy Hatcher.

This patch sends the data to be added or removed from storage over to the UI process. Next and
final steps for the Storage API is to adopt the _WKWebExtensionSQLiteStore to manage the data added
with this API. This is being tracked in https://webkit.org/b/267649.

In addition, this patch adds support for browser.session.setAccessLevel. To do this, add a boolean
representing the access level to WebExtensionContextParameters so that the value can be read when
determining to expose the 'session' API.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_cannot_be_forward_declared):
* Source/WebKit/Shared/Extensions/WebExtensionConstants.h:
* Source/WebKit/Shared/Extensions/WebExtensionContext.serialization.in:
* Source/WebKit/Shared/Extensions/WebExtensionContextParameters.h:
* Source/WebKit/Shared/Extensions/WebExtensionContextParameters.serialization.in:
* Source/WebKit/Shared/Extensions/WebExtensionStorageType.h:
Copied from Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIStorage.h.
(WebKit::toAPIPrefixString):
* Source/WebKit/Shared/Extensions/WebExtensionStorageType.serialization.in:
Copied from Source/WebKit/Shared/Extensions/WebExtensionContext.serialization.in.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIStorageCocoa.mm: Added.
(WebKit::WebExtensionContext::storageGet):
(WebKit::WebExtensionContext::storageGetBytesInUse):
(WebKit::WebExtensionContext::storageSet):
(WebKit::WebExtensionContext::storageRemove):
(WebKit::WebExtensionContext::storageClear):
(WebKit::WebExtensionContext::storageSetAccessLevel):
(WebKit::WebExtensionContext::isValidStorageRequest):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::load):
(WebKit::WebExtensionContext::getTab):
(WebKit::WebExtensionContext::isBackgroundWebView):
(WebKit::WebExtensionContext::setSessionStorageAllowedInContentScripts):
Set the access level and send a message to update the value on the WebExtensionContextParameters.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.cpp:
(WebKit::WebExtensionContext::parameters const):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
(WebKit::WebExtensionContext::isSessionStorageAllowedInContentScripts const):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIStorageAreaCocoa.mm:
(WebKit::WebExtensionAPIStorageArea::isPropertyAllowed):
(WebKit::WebExtensionAPIStorageArea::get):
(WebKit::WebExtensionAPIStorageArea::getBytesInUse):
(WebKit::WebExtensionAPIStorageArea::set):
(WebKit::WebExtensionAPIStorageArea::remove):
(WebKit::WebExtensionAPIStorageArea::clear):
(WebKit::WebExtensionAPIStorageArea::setAccessLevel):
(WebKit::WebExtensionAPIStorageArea::onChanged):
(WebKit::WebExtensionAPIStorageArea::quotaBytes):
(WebKit::WebExtensionAPIStorageArea::quotaBytesPerItem):
(WebKit::WebExtensionAPIStorageArea::maxItems):
(WebKit::WebExtensionAPIStorageArea::maxWriteOperationsPerHour):
(WebKit::WebExtensionAPIStorageArea::maxWriteOperationsPerMinute):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIStorageCocoa.mm:
(WebKit::WebExtensionAPIStorage::isPropertyAllowed):
(WebKit::WebExtensionAPIStorage::local):
(WebKit::WebExtensionAPIStorage::session):
(WebKit::WebExtensionAPIStorage::sync):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIStorage.h:
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIStorageArea.h:
(WebKit::WebExtensionAPIStorageArea::WebExtensionAPIStorageArea):
* Source/WebKit/WebProcess/Extensions/Cocoa/WebExtensionContextProxyCocoa.mm:
(WebKit::WebExtensionContextProxy::getOrCreate):
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIStorage.idl:
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIStorageArea.idl:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.cpp:
(WebKit::WebExtensionContextProxy::setStorageAccessLevel):
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.h:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIStorage.mm:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/273286@main
